### PR TITLE
DIP0008: Replace signing attempts with a majority of quorums signing the same height

### PR DESCRIPTION
**Pull request by UdjinM6**
The main idea is to let all active ChainLock quorums create new CLSIG messages at every block and to require the majority of these quorums to vote for the same hash to actually lock the chain. Unlike in current system, having a malicious quorum will no longer be enough to disrupt ChainLocks via withholding attack or by producing "fake" CLSIGs (CLSIGs for some future height with some random hashes). This also deprecates the "signing attempts" part cause it feels redundant in this setup.

PoC for Dash Core: https://github.com/dashpay/dash/pull/4016

### CODE REVIEW

File: dip-0008.md [hunk 1]
```diff
@@ -181,56 +152,101 @@ message is called `CLSIG` and has the following structure:
...
+
+### Handling of blocks signed via a single LLMQ
+
+Upon receipt, each node must perform the following verification before
+announcing it to other nodes:
+
+  1. Based on the deterministic masternode list at the given height,
+  a set of active quorums must be calculated.
+  2. A request id must be calculated.
+  3. A quorum must be selected from the set based on the calculated request id.
```
Comment (xdustinface) [inline dip-0008.md:209]:
```suggestion
  1. Based on the deterministic masternode list at the given height,
  the set of active quorums must be calculated.
  2. The request id must be calculated.
  3. The quorum must be selected from the set based on the calculated request id.
```

Might be only me but `A/a` sounds like there could be multiple but there can only be one of those for the specific height, no?

File: dip-0008.md [hunk 2]
```diff
@@ -181,56 +152,101 @@ message is called `CLSIG` and has the following structure:
...
+
+### Handling of blocks signed via multiple LLMQs
+
+If the `signers` bitset has exactly 1 bit set, each node must perform the
+following verification before announcing it to other nodes:
+
+  1. Based on the deterministic masternode list at the given height,
+  a set of active quorums must be calculated.
+  2. A request id must be calculated using the bit in the `signers` bitset set to `1` as a `step`.
+  3. A `step`th quorum must be selected from the set.
```
Comment (xdustinface) [inline dip-0008.md:222]:
```suggestion
  1. Based on the deterministic masternode list at the given height,
  the set of active quorums must be calculated.
  2. The request id must be calculated using the bit in the `signers` bitset set to `1` as a `step`.
  3. The `step`th quorum must be selected from the set.
```
Like above

File: dip-0008.md [hunk 3]
```diff
@@ -181,56 +152,101 @@ message is called `CLSIG` and has the following structure:
...
+
+When using multiple LLMQs, the node must wait for the majority of LLMQs to sign
+the same block hash and then a new `CLSIG` message with an aggregated `sig` and
+a combined `signers` field should be formed and relayed to other nodes.
+
+If the `signers` bitset has at least `signersSize / 2 + 1` bits set to `1`, each node
+must perform the following verification before announcing it to other nodes:
+
+  1. Based on the deterministic masternode list at the given height,
+  a set of active quorums must be calculated.
```
Comment (xdustinface) [inline dip-0008.md:233]:
```suggestion
  the set of active quorums must be calculated.
```

Comment (xdustinface) [reply dip-0008.md:233]: Same

**Review by xdustinface — CHANGES_REQUESTED** [review]
Good idea 👍 I didn't think that much about the details/impact yet but overall i think it makes sense, see comments below for some thoughts after a first read + [table-of-contents](https://github.com/dashpay/dips/blob/57d4cf215678ad8403fcdf8f935c65138548918d/dip-0008.md#table-of-contents) section also needs some adjustments :) e.g. this dead link -> [signing-attempts](https://github.com/dashpay/dips/blob/57d4cf215678ad8403fcdf8f935c65138548918d/dip-0008.md#signing-attempts)

Also, did you think about just making a separate DIP out of this? I honestly don't have a strong opinion on wether it should be separate or not but i think it could make sense to make a new and refer to this one. Thoughts?

File: dip-0008.md [hunk 4]
```diff
@@ -125,54 +125,25 @@ are InstantSend locked in a retroactive way if they were unknown before appearin
...
-succeeds or fails without performing additional attempts.
-
-The request id is `SHA256("clsig", blockHeight)` and the message hash is the block hash of the
-previously successful attempt.
+## Signing blocks
+
+The request id to sign a block must be calculated deterministically once for
+each `blockHeight`.
+
+The message hash is always the hash of the signed block.
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply dip-0008.md:135]: Signing is done via an interface defined in DIP0007, mentioned this in 2effb4d.

(side note: DIP0007 needs updating...)

File: dip-0008.md [hunk 5]
```diff
@@ -181,56 +152,101 @@ message is called `CLSIG` and has the following structure:
...
-  `SHA256(llmqType, quorumHash, SHA256(height), blockHash)`. `llmqType` and `quorumHash`
-  must be taken from the quorum selected in 1.
+This is an optional method which can be used only by nodes running on protocol
+70219 or newer.
 
+The request id for each LLMQ is calculated as `SHA256("clsig", blockHeight, step)`
+where `step` starts from `0` and increments by `1` starting with the most recent LLMQ.
 
-## Handling of signed blocks
+After a LLMQ member has successfully recovered the final ChainLocks signature
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply dip-0008.md:173]: This part is basically a copy/paste from the original text https://github.com/dashpay/dips/pull/85/files#diff-93e810c1866c9c5a380149d70483041dba93c8b2c274945b9313813a7e4ab10bR153 and I think in this case we refer to ChainLocks as a technology (as defined in `Abstract`), not as an individual object (the object is the signature here).

File: dip-0008.md [hunk 6]
```diff
@@ -17,10 +17,13 @@
 1.  [Prior Work](#prior-work)
 1.  [Used LLMQ type](#used-llmq-type)
 1.  [Safe transactions](#safe-transactions)
-1.  [Signing attempts](#signing-attempts)
-1.  [Finalization of signed blocks](#finalization-of-signed-blocks)
+1.  [Signing blocks](#signing-blocks)
+    1.  [Signing blocks using single LLMQ](#signing-blocks-using-single-llmq)
+    1.  [Signing blocks using multiple LLMQs](#signing-blocks-using-single-llmqs)
```
Comment (xdustinface) [inline dip-0008.md:22]:
```suggestion
    1.  [Signing blocks using multiple LLMQs](#signing-blocks-using-multiple-llmqs)
```

Comment (UdjinM6) [reply dip-0008.md]: ooops, fixed

**Comment by QuantumExplorer** [comment]
I'm not sure I understand the reasoning behind this change. Could you elaborate more @UdjinM6? Taking control of a 400 member quorum requires close to 60% of the network or 50% and a lot of hash power.

File: dip-0008.md [hunk 7]
```diff
@@ -181,56 +157,101 @@ message is called `CLSIG` and has the following structure:
...
-and removed from the currently active chain since a signed block has already
-been received. If the correct block is already present locally, its chain
-should be activated as the new active chain. If the correct block is not known
-locally, it must wait for this block to arrive and request it from other nodes
-if necessary.
+This message is propagated through the inventory system.
+
+## Handling of signed blocks
+
+The node should ensure that the fully singed block is locally accepted as the
```
Comment (QuantumExplorer) [inline dip-0008.md:192]: signed

**Review by xdustinface — CHANGES_REQUESTED** [review]
For me the `step` stuff looks a bit confusing tbh, thoughts about c0648013f3a3daf93bf67f36d41ea41d8fa6b9af? See xdustinface/dash@36628cf40f392b4f204dab239ac33a85064b2b38 for an implementation of it.

**Comment by UdjinM6** [comment]
@xdustinface Makes sense, picked both. Thanks! 👍

**Review by xdustinface — CHANGES_REQUESTED** [review]
Some more (and my last) points: 
- Let's maybe just move the signer stuff into the table like in 3f02341?
- Maybe we should mention somewhere that the value of `SPORK_19_CHAINLOCKS_ENABLED` is used to decide if multi LLMQ signing is enabled or not. 
- It looks like there is a mismatch between dashpay/dash#4016 and the changes here. For me when i read it here it sounds like:
     - If single LLMQ signing is enabled, i.e. the spork is active and != 1, the `CLSIG` has structure `A` (signers fields are not present)
     - If multi LLMQ signing is enabled, i.e. the spork value is 1, the `CLSIG` has structure `B` (signer fields present).

  But it's actually always structure `B` with protocol version `70220` but with signers size 0, no? So we could maybe merge those both tables, put them somewhere (maybe a separate message section) below the signing section and somehow mark the new fields introduced by the new version. Thoughts?

**Review by xdustinface — APPROVED** [review]
Looks all good to me now 👍

File: dip-0008.md [hunk 8]
```diff
@@ -125,112 +129,125 @@ are InstantSend locked in a retroactive way if they were unknown before appearin
...
-the [DIP007 `GetMostSignedSession` operation](https://github.com/dashpay/dips/blob/master/dip-0007.md#uint256-getmostsignedsessionuint256-id-int-activellmqs), which is the hash of the block
-which had the most signatures in the last attempt. After a few attempts, a
-request should result in a recovered threshold signature which indicates
-consensus has been reached.
+The sign hash for the BLS signature is calculated as
+`SHA256(llmqType, quorumHash, requestId, blockHash)`, where `llmqType` and
+`quorumHash` are taken from the corresponding LLMQ.
 
-## Finalization of signed blocks
+There are two possible ways to form a ChainLock: using single LLMQ and using
```
Comment (thephez) [inline dip-0008.md:158]:
```suggestion
There are two possible ways to form a ChainLock: using a single LLMQ and using
```

File: dip-0008.md [hunk 9]
```diff
@@ -125,112 +129,125 @@ are InstantSend locked in a retroactive way if they were unknown before appearin
...
-and removed from the currently active chain since a signed block has already
-been received. If the correct block is already present locally, its chain
-should be activated as the new active chain. If the correct block is not known
-locally, it must wait for this block to arrive and request it from other nodes
-if necessary.
+The node should ensure that the fully signed block is locally accepted as the
+next block. If the correct block is already present locally, its chain should
+be activated as the new active chain tip. If the correct block is not known
+locally, the node must wait for this block to arrive and request it from other
+nodes if necessary. If an alternative block for the same height is received,
```
Comment (thephez) [inline dip-0008.md:189]: Suggested rewording of this sentence:

If the correct block is already present locally and an alternative block for the same height is received, the alternative block must not be added to the currently active chain.

**Comment by thephez** [comment]
Is the "Calculations" section still accurate with this change? Also, I was thinking it might be nice to include a link to the original version of the document (whatever the most recent change was) so people can easily refer back to it. I think that could be added to the "Prior Work" section.

File: dip-0008.md [hunk 10]
```diff
@@ -125,112 +129,125 @@ are InstantSend locked in a retroactive way if they were unknown before appearin
...
+it must not be added to the currently active chain, if the correct block is
+already present locally.
```
Comment (UdjinM6) [inline dip-0008.md:191]:
```suggestion
nodes if necessary. If the correct block is already present locally and an alternative block for the same height is received, the alternative block must not be added to the currently active chain.
```

**Comment by UdjinM6** [comment]
> Is the "Calculations" section still accurate with this change?

Should be accurate for a single LLMQ mode but yeah, for a multi-quorum mode you'd need to disrupt or control 2 quorums at the same time so I guess you should take all these probabilities roughly to the power of 2 or so? Let's ping @DarrenTa :)

> Also, I was thinking it might be nice to include a link to the original version of the document (whatever the most recent change was) so people can easily refer back to it. I think that could be added to the "Prior Work" section.

Hmm... current version in master is https://github.com/dashpay/dips/blob/81e81b52799114191729364ed36dc0e372062fbe/dip-0008.md but I'm not sure if it makes a lot of sense to link back to itself, especially in a git repo..

**Comment by DarrenTa** [comment]
Thank you so much for the ping.  I can provide the following:

1)  I can review this change and form an opinion, backed by data, if this should be live or not.
2)  I do agree with @QuantumExplorer above.  I would want a comparable security level out of the new process.  
3)  I almost wonder if a **straight up multisig** would be the way to go?  The threshold setup does require overhead.
4)  In response to @thephez perhaps DIP0008 should not be edited at all.  It would be best to make a new DIP and then change the status of 0008 (to something like superseded).

I have cited DIP0008 in my academic work, it would make my citations cleaner if DIP0008 was not changed significantly.  I was lazy and did not put date accessed.

I've also cited (withdrawn) 0005 and think the citations provide for a robust discussion, even if not live.  No reason 0008 has to be edited to agree with the live implementation.

I'll circle back here in a week and report back with a proper analysis.

**Comment by UdjinM6** [comment]
Hmm.. ok, you convinced me to at least create an alternative PR to have smth to compare to :) Pls see #89.

**Comment by UdjinM6** [comment]
closing in fav of #89

**Pull request closed without merging**